### PR TITLE
Migrate top-level connection

### LINKED ISSUE #4158: The `@relayId` fields cannot be aliased when queried using the Connection API.
The `@relayId` fields cannot be aliased when queried using the Connection API. 

Given the following typeDefs:
```graphql
type Movie {
    dbId: ID! @id @unique @relayId
    title: String!
    genre: Genre! @relationship(type: "HAS_GENRE", direction: OUT)
    actors: [Actor!]! @relationship(type: "ACTED_IN", direction: OUT)
}

type Genre {
    dbId: ID! @id @unique @relayId
    name: String!
}

type Actor {
    dbId: ID! @id @unique @relayId
    name: String!
}
```
And the following query:
```graphql
query {
    movies {
        testAliasID: id
        testAliasDBID: dbId
        title
        genre {
            testAliasID: id
            testAliasDBID: dbId
            name
        }
        actors {
            testAliasID: id
            testAliasDBID: dbId
            name
        }       
    }
}
```

It's possible to obtain both the `id` and `dbId` fields with their respective  GraphQL alias.

However when queried using the Connection API:
```graphql
query {
    moviesConnection {
        totalCount
        edges {
            node {
                testAliasID: id
                testAliasDBID: dbId
                title
                genre {
                    testAliasID: id
                    testAliasDBID: dbId
                    name
                }
                actors {
                    testAliasID: id
                    testAliasDBID: dbId
                    name
                }
            }
        }
    }
}
```

The following error appears: `GraphQLError: Cannot read properties of undefined (reading 'toString')`

### PR DESCRIPTION
**Pull request by MacondoExpress**
# Description

Migrate the top-level connection to the new translation layer

## Complexity

Complexity: Med

# Issue

Closes #4158 

# Checklist

The following requirements should have been met (depending on the changes in the branch):

- [x] Documentation has been updated
- [x] TCK tests have been updated
- [x] Integration tests have been updated
- [x] Example applications have been updated
- [x] New files have copyright header
- [x] CLA (https://neo4j.com/developer/cla/) has been signed

### CODE REVIEW

**Comment by neo4j-team-graphql** [comment]
Made some changes to the Neo4j GraphQL Toolbox? Or made some changes to the core library perhaps?

Check out your changes at https://neo4j-graphql-toolbox-4133.surge.sh

Use for instance a local dbms hosted on [Neo4j Desktop](https://neo4j.com/download/) or a cloud-hosted [Aura Free](https://neo4j.com/cloud/aura-free/) dbms.

**Comment by neo4j-team-graphql** [comment]
# Performance Report

| name | dbHits | old dbHits | time (ms) | old time (ms) | maxRows |
| ---- | ------ | ---------- | --------- | ------------- | ------- |
| 🟥 connections.ConnectionWithSortAndCypher | 29203 | 16325 | 145 | 197 | 2174 |


🟥 - Performance worsened (dbHits)
🟩 - Performance improved (dbHits)
🟦 - New test
        

<details>
    <summary>Show Full Table</summary>
    
| name | dbHits | old dbHits | time (ms) | old time (ms) | maxRows |
| ---- | ------ | ---------- | --------- | ------------- | ------- |
|  aggregations.TopLevelAggregate | 3403 | 3403 | 63 | 60 | 1134 |
|  aggregations.NestedAggregation | 16553 | 16553 | 103 | 117 | 2174 |
|  aggregations.AggregationWithWhere | 11979 | 11979 | 77 | 77 | 2174 |
|  aggregations.AggregationWhereWithinNestedRelationships | 22116987 | 22116987 | 4742 | 5161 | 2008534 |
|  aggregations.AggregationWhereWithinNestedConnections | 22116987 | 22116987 | 4465 | 4178 | 2008534 |
|  aggregations.NestedCountFromMovieToActors | 9734 | 9734 | 58 | 70 | 2174 |
|  aggregations.NestedCountFromActorsToMovie | 9937 | 9937 | 87 | 111 | 2174 |
|  aggregations.DeeplyNestedCount | 11062443 | 11062443 | 6752 | 6639 | 2008534 |
|  batch-create.BatchCreate | 4200 | 4200 | 173 | 249 | 600 |
|  batch-create.BatchCreateSmall | 77 | 77 | 85 | 186 | 11 |
|  connect.createAndConnect | 14424 | 14424 | 247 | 360 | 3003 |
|  connections.Connection | 14082 | 14082 | 80 | 130 | 2174 |
|  connections.NestedConnection | 41118 | 41118 | 153 | 176 | 4516 |
|  connections.ConnectionWithSort | 3284 | 3284 | 95 | 139 | 1040 |
| 🟥 connections.ConnectionWithSortAndCypher | 29203 | 16325 | 145 | 197 | 2174 |
|  create.SimpleMutation | 7 | 7 | 111 | 120 | 1 |
|  cypher-directive.TopLevelMutationDirective | 1135 | 1135 | 39 | 63 | 1134 |
|  delete.SimpleDelete | 19401 | 19401 | 405 | 375 | 1040 |
|  delete.NestedDeleteInUpdate | 19542 | 19542 | 336 | 346 | 1183 |
|  2871.NestedRelationshipFilter | 23048 | 23144 | 93 | 113 | 4411 |
|  2925.SingleRelationshipFilter | 6468 | 6468 | 68 | 71 | 1040 |
|  2925.NestedSingleRelationshipFilter | 22900 | 22900 | 122 | 198 | 2174 |
|  2925.SingleRelationshipRequiredFilter | 5201 | 5201 | 58 | 77 | 1040 |
|  2925.NestedSingleRelationshipRequiredFilter | 9361 | 9361 | 87 | 130 | 1040 |
|  query.SimpleQuery | 3121 | 3121 | 38 | 41 | 1040 |
|  query.SimpleQueryWithRelationship | 16162 | 16162 | 63 | 88 | 2174 |
|  query.QueryWhere | 9705 | 9722 | 53 | 71 | 2163 |
|  query.SimpleQueryWithNestedWhere | 9883 | 9900 | 79 | 100 | 2163 |
|  query.Nested | 10096041 | 10096041 | 12737 | 13837 | 2008534 |
|  query.NestedWithFilter | 10074401 | 10074401 | 12489 | 13129 | 2004000 |
|  query.OrFilterOnRelationships | 43016 | 42102 | 290 | 316 | 1845 |
|  query.OrFilterOnRelationshipsAndNested | 36737 | 35337 | 360 | 382 | 1845 |
|  query.QueryWithNestedIn | 13286 | 13492 | 99 | 120 | 1178 |
|  query.NestedConnectionWhere | 9834 | 9834 | 94 | 128 | 2174 |
|  query.DeeplyNestedConnectionWhere | 9965 | 9853 | 142 | 141 | 2174 |
|  query.DeeplyNestedWithRelationshipFilters | 19212 | 19086 | 198 | 292 | 1620 |
|  query.NestedWithRelationshipSingleFilters | 3881 | 3881 | 280 | 370 | 1134 |
|  query.Fulltext | 80 | 80 | 48 | 71 | 16 |
|  query.FulltextWithNestedQuery | 587 | 587 | 66 | 68 | 84 |
|  sorting.SortMultipleTypes | 3515 | 3515 | 95 | 131 | 1040 |
|  sorting.SortMultipleTypesWithCypherWithCypher | 14525 | 14502 | 139 | 150 | 2174 |
|  sorting.SortOnNestedFields | 14082 | 14082 | 73 | 80 | 2174 |
|  sorting.SortDeeplyNestedFields | 43198 | 43198 | 125 | 163 | 4516 |
|  sorting.SortWithTopLevelCypher | 3396 | 3396 | 90 | 115 | 1040 |
|  unions.SimpleUnionQuery | 321 | 321 | 120 | 93 | 35 |
|  unions.SimpleUnionQueryWithMissingFields | 293 | 293 | 99 | 94 | 35 |
|  unions.NestedUnion | 410637 | 410637 | 466 | 508 | 33033 |
|  unions.NestedUnionWithMissingFields | 384611 | 384611 | 415 | 506 | 33033 |
|  update.NestedUpdate | 16143 | 16143 | 157 | 190 | 2002 |

</details>
        
Old Schema Generation: 1:03.492 (m:ss.mmm)
Schema Generation: 1:04.587 (m:ss.mmm)
Old Subgraph Schema Generation: 1:23.021 (m:ss.mmm)
Subgraph Schema Generation: 1:24.673 (m:ss.mmm)

File: packages/graphql/src/translate/queryAST/factory/OperationFactory.ts [hunk 1]
```diff
@@ -72,17 +73,48 @@ export class OperationsFactory {
...
+                op.nodeAlias = TOP_LEVEL_NODE_NAME;
+                return op;
+            }
+            throw new Error("Not implemented");
         }
-
-        const op = this.createReadOperation(entity, resolveTree, context) as ReadOperation;
-        op.nodeAlias = TOP_LEVEL_NODE_NAME;
-        return op;
+        return this.createReadOperation(entity, resolveTree, context);
```
Comment (a-alle) [inline packages/graphql/src/translate/queryAST/factory/OperationFactory.ts:97]:
```suggestion
if (!isConcreteEntity(entity)) {
        return this.createReadOperation(entity, resolveTree, context);
    }
    const operationMatch = parseOperationField(resolveTree.name, entity);
    if (operationMatch.isCreate) {
        return this.createCreateOperation(entity, resolveTree, context);
    }  
    if (operationMatch.isRead) {
        const op = this.createReadOperation(entity, resolveTree, context) as ReadOperation;
        op.nodeAlias = TOP_LEVEL_NODE_NAME;
        return op;
    }  
    if (operationMatch.isConnection) {
        const topLevelConnectionResolveTree = this.fixResolveTreeForTopLevelConnection(resolveTree);
        const op = this.createConnectionOperationAST(
            entity,
            topLevelConnectionResolveTree,
            context
        ) as ConnectionReadOperation;
        op.nodeAlias = TOP_LEVEL_NODE_NAME;
        return op;
    }
    throw new Error("Not implemented");  
```

Comment (a-alle) [reply packages/graphql/src/translate/queryAST/factory/OperationFactory.ts:97]: this nested if check doesn't look very easy to understand. We can return early and remove the `else` as each if-case returns. We could also consider using a switch-case to understand the set of all possibilities?

File: packages/graphql/src/translate/queryAST/factory/OperationFactory.ts [hunk 2]
```diff
@@ -229,76 +261,141 @@ export class OperationsFactory {
...
+                    resolveTree,
+                    context,
+                    operation,
+                    whereArgs: resolveTreeWhere,
                 });
-                // nodeWhere with the shared filters applied
-                const concreteNodeWhere = getConcreteWhere(target, concreteEntity, nodeWhere);
-                let whereArgs: Record<string, any>;
+            } else {
+                let concreteConnectionOperations: ConnectionReadOperation[] = [];
```
Comment (a-alle) [inline packages/graphql/src/translate/queryAST/factory/OperationFactory.ts:294]: could this binding be a constant, declared at the same time as it's defined?

File: packages/graphql/src/translate/queryAST/factory/OperationFactory.ts [hunk 3]
```diff
@@ -229,76 +261,141 @@ export class OperationsFactory {
...
 
-        if (isConcreteEntity(target)) {
-            checkEntityAuthentication({
-                entity: target.entity,
-                targetOperations: ["READ"],
-                context,
-            });
-
-            const operation = new ConnectionReadOperation({ relationship, directed, target });
+        if (relationship) {
```
Comment (a-alle) [inline packages/graphql/src/translate/queryAST/factory/OperationFactory.ts:273]: Could we extract this if and its nested if-checks into separate functions? This is really hard to follow..

**Comment by MacondoExpress** [comment]
> This seems like a more than coincidental performance degradation - but the Cypher doesn't look noticeable different. Do we have a performance concern in translation here?

We discussed and tested specifically yesterday, even with the higher `dbHits` the response time seems always better with this implementation.  The slightly different `cypher` is caused by a previous inconsistent resolver strategy regarding the `@cypher` fields with sorting fields between the `simple` and `Connection` API. Now both implementations are using the same strategy.

**Comment by darrellwarde** [comment]
> > This seems like a more than coincidental performance degradation - but the Cypher doesn't look noticeable different. Do we have a performance concern in translation here?
> 
> We discussed and tested specifically yesterday, even with the higher `dbHits` the response time seems always better with this implementation. The slightly different `cypher` is caused by a previous inconsistent resolver strategy regarding the `@cypher` fields with sorting fields between the `simple` and `Connection` API. Now both implementations are using the same strategy.

I fear that this isn't the case. If you look at the `ConnectionWithSortAndCypher` performance test, you will see there is a `limit` argument. If you remove this argument, you will see that the DB hits are the same as what the test is showing here on this branch. Whilst we may not be seeing the impacts in this dataset, this will result in exponentially poor performance in a larger dataset, and if the selection set includes nested relationships. The TCK tests for #1364 only show ordering being moved, but the issue is that limiting follows that. So we have moved limiting to after relationship projection, which means that _all_ of the relationships are projected before we apply the limit. It's important that limiting is done as early as practicably possible, which was the case before this change was made. Let's discuss tomorrow!

**Comment by mjfwebb** [comment]
@MacondoExpress changeset? :)

**Review by a-alle — APPROVED** [review]
